### PR TITLE
Update benchmarks & add classgroups

### PR DESCRIPTION
**Pull request by goldenMetteyya**
Hello @dignifiedquire ,

Great Work here! I updated the benchmarks to use criterion so it is more informative. Also I see that this implementation uses RSA, there is a class group implementation have you seen this?

https://github.com/poanetwork/vdf/tree/master/classgroup

I think a trait for the core operations would be nice and then implemented for both backends.

May you look into this?

Also I dont see any serde support so I will integrate that for the structure.

thanks

Efe

### CODE REVIEW

**Comment by dignifiedquire** [comment]
Hey

> Great Work here! 

thank you!

> I updated the benchmarks to use criterion so it is more informative.

Sounds like a good plan

> Also I see that this implementation uses RSA, there is a class group implementation have you seen this?
> 
> [poanetwork/vdf:classgroup@`master`](https://github.com/poanetwork/vdf/tree/master/classgroup)

I hadn't seen this work no. Thanks for pointing it out.
 
> I think a trait for the core operations would be nice and then implemented for both backends.
> 
> May you look into this?

I am not sure this needs a trait, I would say just having a `from_bytes_be` and `to_bytes_be` method on the accumulators and VCs, should be enough, such that it can import from any setup. And then I would pull out the current rsa setup into a freestanding call to do so. What do you think? 
 
> Also I dont see any serde support so I will integrate that for the structure.

What is your use case for this? What structures exactly do you need to serialize from and to? I think adding serde make sense, but only behind a feature gate. 


There is a lot of code by now on this branch, which I can't merge as is. What's your current plan with this branch?

**Comment by goldenMetteyya** [comment]
> Hey
> 
> > Great Work here!
> 
> thank you!
> 
> > I updated the benchmarks to use criterion so it is more informative.
> 
> Sounds like a good plan
> 
> > Also I see that this implementation uses RSA, there is a class group implementation have you seen this?
> > [poanetwork/vdf:classgroup@`master`](https://github.com/poanetwork/vdf/tree/master/classgroup)
> 
> I hadn't seen this work no. Thanks for pointing it out.
Yes they use gmp lib which encodes the integers as big endian twos compliment so it can be serialized into a BigInt library from rust that is used in this crate. 
> 
> > I think a trait for the core operations would be nice and then implemented for both backends.
> > May you look into this?
> 
> I am not sure this needs a trait, I would say just having a `from_bytes_be` and `to_bytes_be` method on the accumulators and VCs, should be enough, such that it can import from any setup. And then I would pull out the current rsa setup into a freestanding call to do so. What do you think?

I think the simpler the Better always. When using class groups over RSA is the only part that is changed how the n (modulus) is derived as the prime? I am little confused about that part.
> 
> > Also I dont see any serde support so I will integrate that for the structure.
> 
> What is your use case for this? What structures exactly do you need to serialize from and to? I think adding serde make sense, but only behind a feature gate.

Yes I think a feature gate is a good idea. For me its about serializing to disk and over the wire. 
> 
> There is a lot of code by now on this branch, which I can't merge as is. What's your current plan with this branch?

Sorry about that I didn't see that my main was this PR and it got many commits. I am cleaning up the master branch so that it can be easier to merge, will write when think its there. One of the main things I did was bring some of the math functions from the RSA crate to remove deps as much as possible which I like, do you not like this?

**Comment by dignifiedquire** [comment]
> One of the main things I did was bring some of the math functions from the RSA crate to remove deps as much as possible which I like, do you not like this?

I want to move those math functions into [my fork of num_bigint](https://github.com/dignifiedquire/num-bigint/) instead, and then use them in the RSA implementation and here, as I want to use that as the baseline for all the math.

**Comment by goldenMetteyya** [comment]
Hey I cleaned up a little, in regards to having multiple backends I think having a feature flag is smart, but you still think that the trait isn't the best route? I have one simple PrimeGroup trait.

File: src/group/mod.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+
+//pub mod classygroup;
```
Comment (dignifiedquire) [inline src/group/mod.rs]: why is this commented out?

File: src/lib.rs [hunk 2]
```diff
@@ -1,11 +1,45 @@
...
+extern crate serde; 
+extern crate failure;
+#[macro_use] extern crate failure_derive;
+extern crate num_iter;
+extern crate num_bigint;
+extern crate num_integer;
+extern crate num_traits;
+extern crate rand;
+extern crate byteorder;
+extern crate blake2;
```
Comment (dignifiedquire) [inline src/lib.rs]: Please remove all the `extern crate`, this is using 2018 edition, where they are not needed anymore.

File: src/lib.rs [hunk 3]
```diff
@@ -1,11 +1,45 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
 #![deny(clippy::all, clippy::perf, clippy::correctness)]
 #![allow(clippy::unreadable_literal, clippy::many_single_char_names)]
 #![warn(clippy::type_complexity, clippy::too_many_arguments)]
+#![feature(uniform_paths)]
```
Comment (dignifiedquire) [inline src/lib.rs]: I would like to keep compiling to stable, why is this required?

**Comment by goldenMetteyya** [comment]
Hey @dignifiedquire ,

So I addressed most of your points. The thing with classgroups is I'm still finalizing the use. I understand how to sample the prime but compared to RSA where you return the exponent I guess its the generator but the corresponding in the class group I have been confused about.

File: Cargo.toml [hunk 4]
```diff
@@ -6,17 +6,37 @@ edition = "2018"
...
+rand = "0.6"
+failure = "0.1"
+failure_derive = "0.1"
 blake2 = "0.8.0"
 generic-array = "0.12.0"
 byteorder = "1.2.7"
 bitvec = "0.9.0"
+serde = "1.0"
+serde_derive = "1.0"
+classgroup = { git = "https://github.com/stichtingorganism/classygroup" }
```
Comment (dignifiedquire) [inline Cargo.toml]: this should be optional, otherwise it gets compiled even if the `class_group` feature is not active

File: README.md [hunk 5]
```diff
@@ -4,6 +4,29 @@
...
+Extra Readings:
+
+    - https://kodu.ut.ee/~lipmaa/papers/lip12b/cl-accum.pdf
+    - https://www.cs.purdue.edu/homes/ninghui/papers/accumulator_acns07.pdf
+    - https://github.com/Chia-Network/vdf-competition/blob/master/classgroups.pdf
+    - https://eprint.iacr.org/2018/623.pdf
+
+## Usage
+
+- Install the [GNU Multiple Precision Library](https://gmplib.org/)
```
Comment (dignifiedquire) [inline README.md]: can you add a note that this is only required when using the classgroup feature?

File: benches/accumulators_benchmarks.rs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+extern crate num_traits;
+extern crate rand;
+extern crate rand_chacha;
+#[macro_use]
+extern crate criterion;
+extern crate blake2;
+
+use criterion::Criterion;
+
+//These benches are taken from various places that the subcomnets were brought into this crate
```
Comment (dignifiedquire) [inline benches/accumulators_benchmarks.rs]: typo: subcomnets

File: src/traits.rs [hunk 7]
```diff
@@ -130,3 +138,28 @@ pub trait DynamicVectorCommitment: StaticVectorCommitment {
...
+/// The order of the class group of a negative prime discriminant d, where |d| ≡ 3 mod 4,
+/// is believed to be difficult to compute when |d| is sufficiently large, making the order
+/// of the class group effectively unknown. Therefore, a suitable discriminant — and its associated
+/// class group — can be chosen without the need for a trusted setup, which is a major advantage for
+/// using class groups in applications requiring groups of unknown order.
+pub trait PrimeGroup {
+    #[cfg(feature = "class_group")]
+    /// Generates the Prime elements from the group that is used via a seed
+    /// Returns first the prime and second the generator used
+    fn generate_primes(seed: &[u8], int_size_bits: usize) -> Result<(BigUint, BigUint), Error>;
```
Comment (dignifiedquire) [inline src/traits.rs]: This seems to be in conflict with the impl you did and this means that we can't use both types of accumulators at the same time, as these will be conflicting

Comment (goldenMetteyya) [reply src/traits.rs]: Yes I have been looking at the issue with why I am confused, the class group library I am using exposes essential method initialize a group from a discriminator that is just a seed and it has a method that generates the negative prime needed for the class group. then it can use that to initialize the group from two generator. This yields a structure that encapsulates the elements needed for the class group. You have a common reference string for that group so it needs a seed but the RSA doesn't have that. There is a method that generates a generator_for_discriminant() for that group. I am figuring out the right way to use for the accumulator. But I believe there is an abstraction that can work. However in practice you don't want a random seed but a known CSR for the group init. For now I removed it.

Comment (goldenMetteyya) [reply src/traits.rs]: Or instead of abstracting the group from the accumulator there are already traits for that and have two implementations but from my understanding there would be a lot of duplication.

**Comment by goldenMetteyya** [comment]
> @goldenMetteyya CI should be working now when you push again. I have fixed some small things but seems there are some more issues on compiling classgroups, you can see those here: #2

sorry I had fixed then last night locally but for got to push the changes.

**Comment by goldenMetteyya** [comment]
@dignifiedquire I removed all the code that is directly related to class groups and moved to the crate  to isolate it from here and make it more modular.

**Comment by dignifiedquire** [comment]
great, this is looking good, I would merge it now and once you get the classgroup integration working you can make a new pr uncommenting it. sound good?
